### PR TITLE
Update config read order

### PR DESCRIPTION
**Pull request by ryanrath**
## Description
Updating the order in which `Configuration` classes read / process their local config files, if present, to be in alphabetical order.

## Motivation and Context
It's desirable to have these files processed in some deterministic manner as opposed to however the FS decides to return them.

## Tests performed
Unit test added that ensures the local config files are processed in alphabetical order.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] My code follows the code style of this project as found in the **CONTRIBUTING** document.
- [X] I have added tests to cover my changes.
- [X] All new and existing tests passed.

### CODE REVIEW

File: tests/artifacts/xdmod-test-artifacts/xdmod/configuration/input/read_order.json [hunk 1]
```diff
@@ -0,0 +1,9 @@
+[
+  [
+    {
+      "base_file": "roles",
+      "local_dir": ["roles.d", "read_order"],
+      "expected": "roles-read_order"
+    }
+  ]
+]
```
Comment (smgallo) [inline tests/artifacts/xdmod-test-artifacts/xdmod/configuration/input/read_order.json]: Can we add newlines at the end of these 3 files? I know if I open them in vim it's going to add a newline by default.

File: tests/artifacts/xdmod-test-artifacts/xdmod/configuration/output/roles-read_order.json [hunk 2]
```diff
@@ -0,0 +1 @@
...egend_type":"right_center","limit":20,"show_filters":true,"start":0,"timeseries":true,"title":"Total CPU Hours by Job Size"},{"data_series":{"data":[{"combine_type":"side","display_type":"column","filters":{"data":[],"total":0},"group_by":"none","has_std_err":true,"id":3.0e-14,"ignore_global":false,"log_scale":false,"long_legend":true,"metric":"avg_processors","realm":"Jobs","sort_type":"none","std_err":true,"value_labels":false,"x_axis":false}],"total":1},"global_filters":{"data":[],"total":0},"legend_type":"off","limit":20,"show_filters":true,"start":0,"timeseries":true,"title":"Avg Job Size (Core Count)"},{"data_series":{"data":[{"combine_type":"side","display_type":"pie","filters":{"data":[],"total":0},"group_by":"pi","has_std_err":true,"id":4.0e-14,"ignore_global":false,"log_scale":false,"long_legend":true,"metric":"total_cpu_hours","realm":"Jobs","sort_type":"value_desc","std_err":true,"value_labels":true,"x_axis":false}],"total":1},"global_filters":{"data":[],"total":0},"legend_type":"off","limit":10,"show_filters":true,"start":0,"timeseries":false,"title":"Total CPU Hours by PI"}]},"mgr":{"display":"Manager","type":"feature"},"dev":{"display":"Developer","type":"feature"}}}
```
Comment (plessbd) [inline tests/artifacts/xdmod-test-artifacts/xdmod/configuration/output/roles-read_order.json]: please pretty print this for easier diffing in the future.

File: classes/Configuration/Configuration.php [hunk 3]
```diff
@@ -322,15 +321,26 @@ public function initialize($force = false)
                 }
 
                 $fullPath = $this->localConfigDir . "/" . $file;
+                $files[] = $fullPath;
```
Comment (jpwhite4) [inline classes/Configuration/Configuration.php]: no need for a local scope variable here. You can add the string directly to the array:
```
$files[] = $this->localConfigDir . "/" . $file;
```

File: classes/Configuration/Configuration.php [hunk 4]
```diff
@@ -322,15 +321,26 @@ public function initialize($force = false)
...
+
+            }  //  while ( false !== ( $file = readdir($dh) ) )
+
+            // Sort the retrieved .json files.
+            sort($files, SORT_STRING);
```
Comment (jpwhite4) [inline classes/Configuration/Configuration.php]:
```suggestion
            sort($files, SORT_LOCALE_STRING);
```

File: classes/Rest/Controllers/WarehouseControllerProvider.php [hunk 5]
```diff
@@ -1401,18 +1401,18 @@ private function getJobData(Application $app, XDUser $user, $realm, $jobId, $act
...
     {
-        $configFile = new XdmodConfiguration(
-            'rawstatistics.json',
-            CONFIG_DIR,
-            null,
-            array(
-                'local_config_dir' => implode(DIRECTORY_SEPARATOR, array(CONFIG_DIR, 'rawstatistics.d'))
+        $rawstats = XdmodConfiguration::assocArrayFactory('rawstatistics.json', CONFIG_DIR);
+
+        $realmExists = count(
```
Comment (jpwhite4) [inline classes/Rest/Controllers/WarehouseControllerProvider.php]: Disadvantage of this code is that you always build an array with all of the realms in it. Even though the code can stop looking when a match is found. Also there is no need to create a temporary array.

```suggestion
        $realmExists = false;
        foreach ($rawstats as $item) {
            if ($item['name'] === $realm) {
                $realmExists = true;
                break;
            }
        }
```

```

**Pull request merged**